### PR TITLE
802.15.4 Support for the NRF52840

### PR DESCRIPTION
**Pull request by dcellucci**
### Pull Request Overview

This pull request adds 802.15.4 support for the NRF52840. 

### Testing Strategy

This pull request was tested by running the Radio_RX and Radio_TX on separate BT840 modules and verifying packet reception.

### TODO or Help Wanted

- [ ] Handling of receive/send buffers is sloppy
- [ ] add external timer support (similar to how the BLE advertising driver works)
- [x] CCA functionality (if the channel is busy right now, it spams CCA until it isn't. This is not standard 802.15.4)
- [ ] ACK functionality
- [ ] CRC functionality

Also, the kernel still needs a unified interface that accounts for multiprotocol and external module radios (it's breaking those right now). 

I'm suggesting something similar to MuxMAC that has a bunch of radio users that queue messages to be sent. 

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by dcellucci** [comment]
Here's a diagram of what I'm thinking the radio mux might do

https://docs.google.com/drawings/d/1KETbaQ3CXbzfTPnK_o22GhJNf6SGBfVR_8t4DEKwbvI/edit?usp=sharing

**Comment by alevy** [comment]
@dcellucci Awesome work!

It looks like there is only one change that's not isolated to the NRF52---not bubbling up the tx_buffer all the way up to the driver.

Can you explain why that's necessary/better? If it's motivated by a design change rather than a requirement of the nrf52 in particular, maybe it's worth separating into different PRs for discussion?

Meanwhile, I rebased from master and tested on two nrf52840dks (after reverting the board pin assignments) and it seems to basically work, with the exception that the PAN ID and address seem to arrive (to userland anyway) incorrectly. But this could also be an artifact of not doing a particularly thorough test.

**Comment by dcellucci** [comment]
@alevy thanks! 

Yes that's part of what i meant by the receive/send buffers are sloppy right now. Currently, the hardware driver copies the contents of the buffer from the client to the `PAYLOAD` array char by char (using the method `replace_radio_buffer`). This functionality I took from the original nrf52 code.

But essentially The format of the `PAYLOAD` array that needs to be sent to the radio is as follows:

     | S0 | LEN | S1 | Buffer |

`S0`, `LEN`, and `S1` are part of the 840's radio (also I have no idea what `S0` and `S1` are for) and should probably not be a part of the higher layers of the stack. Ideally, the higher layers only care about buffer.

The problem was, I was passing up the entire `PAYLOAD` array instead of just the slice that the higher layers care about, so when it replaced the TX buffer at the driver layer with this `PAYLOAD` buffer and tried to send again, it would get a overflow because it was trying to add a header onto an array that already had a header. 

I could try to pass up only a slice of `PAYLOAD` up to the driver, which might fix it. This is also the reason the test isn't working, because I guess i didn't commit the change where I properly offset the PAYLOAD buffer population in `replace_radio_buffer`. 

On a deeper note, however, does this byte by byte population of the hardware buffer that goes on in `replace_radio_buffer`, and the passing of that hardware buffer pointer back up to the driver, jive with the larger philosophy of TockOS? Seems very hacky-C, but I also don't know a good alternative. 

My solution to the latter problem was to make a TakeCell called `kernel_tx` that is the transmit buffer and is taken into the radio when transmit is initiated, and returned when the transmit is completed. No idea if this is the right way to do it though.

**Comment by phil-levis** [comment]
I'd strongly argue that, in the basic packet interface, you want to copy entire packet buffers and not just pass around payloads. There are a lot of reasons for this:

1) Avoids needs for an additional copy; some radios require the entire frame be a single SPI operation
2) Allows other capsules to inspect header fields (through methods); otherwise you need to couple metadata with every frame; this is a requirement for protocols like Thread

**Comment by dcellucci** [comment]
Ah- what I meant by "payload" was more the Nordic radio definition of "payload". That is, the in-RAM representation of the radio packet is as follows
 
Where `PAYLOAD` contains the packet header and the packet payload. As you can see, there are these three additional pieces of data, `S0`, `LENGTH`, and `S1`, whose size is specified in a hardware register and whose order is hardcoded.

It seems like these three values that precede `PAYLOAD` are specific to the NRF52 radio, and therefore shouldn't be part of the higher, hardware-agnostic layers in the stack.

I'm going to try just removing them entirely and see what happens.

**Comment by alevy** [comment]
The high level issue is that the `Radio` HIL is actually already fairly hardware-specific, tailored to the RF233's interface. In particular, I believe the transmit buffer needs to be the PHY payload + two empty bytes at the beginning for the SPI command and the PHY header (a 1 byte, but really 7-bit, length). So that's a bummer.

It seems ideal if we "hack" around the current interface for the NRF52 such that we don't have to address the question of _how_ to redesign the interface to be hardware agnostic. That way we can get NRF52 support off the ground, and set out rethinking the HIL with both the rf233 and nrf52 in mind.

It seems like potentially including the S0 + Length fields in the in-memory representation would work, as long as the S0 field ends up just being ignored.

Does that seem right, or am I missing something (this is from a fairly cursory refresher on the NRF52's interface, and the RF233 driver, so I could totally be wrong).

**Comment by hudson-ayers** [comment]
Just had a chance to look at this, looks like a big step forward, thanks!  A few comments:

1) On filenames and locations - I think it is a little confusing that the file which implements the radio for nrf52840 is just called radio.rs (`chips/nrf52/src/radio.rs`), as this is the same file used to define the kernel HIL for the radio (`kernel/src/hil/radio.rs`) _and_ to define the radio component for Imix (`boards/imix/src/components/radio.rs`), but this file is ultimately a much closer parallel to `capsules/src/rf233.rs`, which is the driver for the rf233 radio used on Imix. Given that, I think we should have a uniform organization of these files -- it seems weird to have the Imix radio driver in capsules and the nrf52 radio driver in chips. Further, while I realize that the nr52840 doesn't AFAIK have a nice name for its multi-protocol radio like RF233, I think a better name for that file would be something like `nrf_radio.rs`

2) I am still not 100% clear on why the requirement for the s0/len/s1 fields means that the rx buffer cant be passed up the stack to the driver. Looking at the rf233 implementation, a 1 byte SPI command and a 1 byte PHY header are required to precede the 802.15.4 frame, which is why `PSDU_OFFSET=2` shows up in `kernel/src/hil/radio.rs`. The actual buffer passed up to/held by the driver contains these additional 2 bytes, but they are skipped when transmitting a frame (`prepare_data_frame() encodes the 15.4 header beginning after the PSDU_OFFSET`) and receiving a frame (`receive()` in `mac.rs` decodes the header
from the received packet by reading in after the PSDU_OFFSET, as does the framer code for reception, and the full buffer is still passed all the way up to the driver, along with the already decoded header and a data_offset which specifies where the actual payload begins, accounting for the PSDU_OFFSET). Given this design, would the easiest fix not be to just make the PSDU_OFFSET configurable based on the underlying hardware, and set it to be the size of s0 + len + s1 for the nrf52? This would allow all other mac packet handling to work exactly as it already does for the rf233 driver, and then we could worry about making the HIL more generic a little later while still having this work now.

Alternatively, of course, if it is in fact possible to simply remove those fields that would be the easiest short term solution.

3) Regarding the proposed RadioMux, that generally seems sensible to me, though I am not sure I can imagine a scenario where we would simultaneously have the MAC and XMAC using the radio, so I think that it is more likely we would just be muxing between MAC and BLE, where Mac is either AwakeMac or XMac.

**Comment by dcellucci** [comment]
@hudson-ayers First I totally agree about the filenames. And from now on I'm going to refer to the `radio.rs` in the `nrf52` folder as `nrf_radio.rs`. I'll make the file names shortly. 

Second (cc @alevy), I have gotten everything working with just having a two-byte header that approximates the RF233 behavior, and I'll push that change shortly. The issue I think is that `driver.rs` and the libtock-c 802154 apps both have implicit assumptions that `PSDU_OFFSET` is 2, and disentangling where these assumptions are in the code I just haven't gotten around to yet. So for now the NRF also has a 2-byte header :) 

Third, I was just going to ask whether a multiplexed BLE/802154 radio was something the community wants.

**Comment by phil-levis** [comment]
The imix radio driver is rf233.rs and in capsules because it's not imix specific; (in theory, untested) any board which has an RF233 can use it. I.e., I could connect an RF233 to a RISC-V chip SPI/interrupt lines and rf233.rs should work.

**Comment by bradjc** [comment]
> Third, I was just going to ask whether a multiplexed BLE/802154 radio was something the community wants.

Yes.

**Comment by dcellucci** [comment]
okay so this new branch (which I had to force, sorry) has an implementation of CCA backoff that's implemented on the hardware level. It means that I have to import/directly access the PPI and TIMER peripherals, as opposed to going through their respective capsules. 

What's the more "Tock" way of doing things? This kind of low-level, potentially dangerous, peripheral access between different HIL modules or a capsule that instead accesses each of these HIL modules indirectly? 

Best, 
DC

**Comment by phil-levis** [comment]
Maybe you and I can do a video chat sometime to talk about an updated Radio HIL as well as these and similar implementation issues?

**Comment by dcellucci** [comment]
Hi All, 

After my conversation with Phil, I fixed the buffer system so it properly utilizes the take cells and buffers (and gets rid of the `unsafe`s). It's now much more similar to the RF233 code. 

One complication is still the use of the DMA pointer. There doesn't seem to be a way to ensure that it's always pointing to something in the DataRAM region (otherwise you get a segfault). I was wondering if you had any suggestions for wrapping that reference so rust makes sure it's doing the right thing.

**Comment by alevy** [comment]
@dcellucci I merged this with master and cleaned up some stylistic issues (e.g. unnecessary use of `unsafe`, stray commented out code and debug statements).

I think this is not ready to merge since it seems to break BLE support, and changes the RadioConfig HIL in such a way that (maybe?) breaks the RF233 implementation.

Where do you think we should go from here? I don't want this work to get lost...

**Comment by dcellucci** [comment]
I'm really sorry- I'll get back on this starting Friday. I was finishing up my thesis...

File: chips/nrf52/src/nrf_radio.rs [hunk 1]
```diff
@@ -588,29 +713,95 @@ impl Radio {
         regs.txpower.set(self.tx_power.get() as u32);
     }
 
-    fn set_dma_ptr(&self) {
+    fn set_dma_ptr(&self, buffer: &'static mut [u8]) -> &'static mut [u8]{
         let regs = &*self.registers;
         unsafe {
-            regs.packetptr.set(PAYLOAD.as_ptr() as u32);
+            regs.packetptr.set(buffer.as_ptr() as u32);
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply chips/nrf52/src/nrf_radio.rs]: Yes, the unsafe block is not needed here!

File: chips/nrf52/src/nrf_radio.rs [hunk 2]
```diff
@@ -588,29 +713,95 @@ impl Radio {
...
-
+        
         if regs.event_ready.is_set(Event::READY) {
             regs.event_ready.write(Event::READY::CLEAR);
             regs.event_end.write(Event::READY::CLEAR);
-            regs.task_start.write(Task::ENABLE::SET);
+            if self.transmitting.get() && regs.state.get() == nrf5x::constants::RADIO_STATE_RXIDLE {
+                if(self.cca_count.get() > 0){
+                    unsafe{
+                        ppi::PPI.disable(ppi::Channel::CH21::SET);
```
[the remark this answers is not in the document]
Comment (dcellucci) [reply chips/nrf52/src/nrf_radio.rs]: I think I looked into this and couldn't find a clean way to do it, but I'm open to suggestions! I need to access a specific Timer channel and a PPI channel. I found that, at least with the timer, going through the timermux didn't give me that level of control.

File: boards/nordic/nrf52dk_base/src/components/radio.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+static mut RADIO_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
+
+// The buffer RF233 packets are received into.
+static mut RADIO_RX_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
+
+// This buffer is used as an intermediate buffer for AES CCM encryption
+// An upper bound on the required size is 3 * BLOCK_SIZE + radio::MAX_BUF_SIZE
+const CRYPT_SIZE: usize = 1;
+static mut CRYPT_BUF: [u8; CRYPT_SIZE] = [0x00; CRYPT_SIZE];
```
Comment (niklasad1) [inline boards/nordic/nrf52dk_base/src/components/radio.rs]: This could be an AtomicUsize?

**Comment by dcellucci** [comment]
Next on my TODOs for this was to implement ACKs, which I'm starting on now.

Regarding the integration into the master branch, how did you all envision it making its way into the main branch? I figure there's three options, in descending order of labor required.

1. consistent radio HIL interface that works for the RF233, Nordic BLE, and Nordic 15.4,
2. interface that separates the RF233 and Nordic Radio, but implements a Mux for each Nordic radio mode.
3. separate interfaces for the Nordic BLE, Nordic 15.4, and RF233 radios

File: chips/nrf52/src/nrf_radio.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alevy) [reply chips/nrf52/src/nrf_radio.rs]: The unsafe isn't actually doing the job we want here. The call to the PPi has no implications for type safety, so should really be marked safe (we make this mistake in other parts of the kernel, so not your fault, but this is the new consensus).

Instead we might want to restrict which modules can enable/disable which specific PPI channels, but that's another question that should be solved differently (e.g. by handing out channels to modules).

**Comment by alevy** [comment]
I think, ideally, there would be different traits for BLE and 15.4, but the rf233 and Nordic 15.4 would implement the same trait --> such that the rest of the IP stack would use them the same.

I think it's fine, for now, to decide at board boot time whether to use the Nordic radio as ble or as 15.4. we can later on imagine have a mux layer that dynamically switches between them based on some policy, but let's not muddle that with this PR.

**Comment by phil-levis** [comment]
I agree with Amit in that it's OK to statically decide "BLE or 802.15.4".

I also agree there should be different traits for 15.4 and BLE. It might be that you add layers on top to provide a uniform abstraction (e.g., IP/6lowpan); but it's important that this doesn't preclude being able to use 15.4 or BLE-specific functionality.

File: boards/nordic/BT840/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+// State for loading and holding applications.
+// How should the kernel respond when a process faults.
+const FAULT_RESPONSE: kernel::procs::FaultResponse = kernel::procs::FaultResponse::Panic;
+
+// Number of concurrent processes this platform supports.
+const NUM_PROCS: usize = 8;
+
+#[link_section = ".app_memory"]
+static mut APP_MEMORY: [u8; 245760] = [0; 245760];
```
Comment (phil-levis) [inline boards/nordic/BT840/src/main.rs]: Since layout.ld is in hex, could you do this in hex as well?

File: capsules/src/rf233.rs [hunk 5]
```diff
@@ -1254,9 +1254,9 @@ impl<S: spi::SpiMasterDevice> radio::RadioConfig for RF233<'a, S> {
         self.cfg_client.set(client);
     }
 
-    fn set_power_client(&self, client: &'static radio::PowerClient) {
+    /*fn set_power_client(&self, client: &'static radio::PowerClient) {
```
Comment (phil-levis) [inline capsules/src/rf233.rs]: Should this be commented out? I think it being commented out will break X-MAC.

File: kernel/src/hil/radio.rs [hunk 6]
```diff
@@ -79,7 +81,7 @@ pub trait RadioConfig {
     fn is_on(&self) -> bool;
     fn busy(&self) -> bool;
 
-    fn set_power_client(&self, client: &'static PowerClient);
+    //fn set_power_client(&self, client: &'static PowerClient);
```
Comment (phil-levis) [inline kernel/src/hil/radio.rs]: I think we need to keep this in the HIL, for low power operation.

File: kernel/src/hil/radio.rs [hunk 7]
```diff
@@ -102,12 +104,81 @@ pub trait RadioConfig {
...
         &self,
-        spi_buf: &'static mut [u8],
-        frame_len: usize,
+        buf: &'static mut [u8],
+        len: usize,
     ) -> (ReturnCode, Option<&'static mut [u8]>);
 }
+
+#[derive(PartialEq, Debug, Copy, Clone)]
+pub enum RadioChannel {
```
Comment (phil-levis) [inline kernel/src/hil/radio.rs]: The enum is good, but I don't think these values should be in the HIL: it looks to be an NRF specific thing? Or are these frequencies (2.4xx MHz)? What do those values mean?

File: chips/nrf52/src/nrf_radio.rs [hunk 8]
```diff
@@ -222,13 +266,24 @@ struct RadioRegisters {
...
     /// Reserved
-    _reserved13: [u32; 3],
+    _reserved15: [u32; 1],
     /// Radio mode configuration register
     /// - Address: 0x650 - 0x654
     modecnf0: ReadWrite<u32, RadioModeConfig::Register>,
     /// Reserved
-    _reserved14: [u32; 618],
+    _reserved16: [u32; 6],
+    /// Clear Channel Assesment (CCA) control register
```
Comment (niklasad1) [inline chips/nrf52/src/nrf_radio.rs]:
```suggestion
    /// Clear Channel Assessment (CCA) control register
```

File: chips/nrf52/src/nrf_radio.rs [hunk 9]
```diff
@@ -544,19 +641,34 @@ impl Radio {
...
-        regs.task_txen.write(Task::ENABLE::SET);
-    }
-
     fn rx(&self) {
         let regs = &*self.registers;
         regs.event_ready.write(Event::READY::CLEAR);
+
+        if self.transmitting.get() {
+            self.tx_buf
+                .replace(self.set_dma_ptr(self.tx_buf.take().unwrap()));
```
Comment (niklasad1) [inline chips/nrf52/src/nrf_radio.rs]: I'm not a fan of these `unwraps` at least change them to expect? 

Could be addressed in another PR

File: chips/nrf52/src/nrf_radio.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (niklasad1) [reply chips/nrf52/src/nrf_radio.rs]: yep, I agree.

Do we _really_ have to mark non-mutating structs as such `static mut PPI`?
Also, I think we should consider to implement `unsafe Sync` for these and snippets as above would work in `safe Rust`

File: chips/nrf52/src/nrf_radio.rs [hunk 10]
```diff
@@ -35,17 +35,42 @@
```
Comment (niklasad1) [inline chips/nrf52/src/nrf_radio.rs]: This change remove `BLE support` for nrf52 do we want that?!

File: kernel/src/hil/radio.rs (hunk 7 above)
Comment (niklasad1) [reply kernel/src/hil/radio.rs]: Those are introduced first as `BLE radio channels` IIRC, i.e, 2400MHz-2480MHz

**Comment by dcellucci** [comment]
Okay I pushed a version that I've confirmed runs both the BLE and 802154 apps individually. Undefined behavior when you run them together. 

Will work on making sure the rf233 apps work again. Though I'll only be able to check if it compiles, not if it runs.

**Comment by dcellucci** [comment]
Not sure why Travis is failing- does anyone have any insight?

I'm interested in getting this PR to the point where you guys feel comfortable merging it, but I'm also trying to add functionality (CRC checks and ACKs specifically) and I have many questions. Should I be pushing those changes here or starting another PR?

**Comment by phil-levis** [comment]
It looks like the code doesn't fit. This means either the chip_layout.ld needs to be updated or we need to see where all of this code space is going. The end of the Travis run:

```
= note: rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 184 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 776 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 1288 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 1728 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3256 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.text' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.ARM.exidx' will not fit in region 'rom': overflowed by 3456 bytes
          rust-lld: error: section '.ARM.exidx' will not fit in region 'rom': overflowed by 3464 bytes
          rust-lld: error: section '.ARM.exidx' will not fit in region 'rom': overflowed by 3472 bytes
          rust-lld: error: section '.ARM.exidx' will not fit in region 'rom': overflowed by 3584 bytes
          rust-lld: error: section '.storage' will not fit in region 'rom': overflowed by 3584 bytes
          rust-lld: error: section '.storage' will not fit in region 'rom': overflowed by 3584 bytes
          rust-lld: error: section '.storage' will not fit in region 'rom': overflowed by 3584 bytes
          rust-lld: error: section '.storage' will not fit in region 'rom': overflowed by 3584 bytes
          rust-lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
```

**Comment by phil-levis** [comment]
I'd suggest separate PRs; most of the core team prefers smaller PRs to review. However, some kind of documentation on where what you're doing is going (e.g., what are the PRs we should expect in the future) would be very helpful.

**Comment by alevy** [comment]
Agree with @phil-levis that a mergeable PR with fewer features is preferable. If nothing else it risks fewer conflict with concurrent work on the BLE stack and 6lowpan.

We should probably indeed figure out where ROM is going, (my guess is much of it is in the 6lowpan stack rather than chip specific stuff since that was the case for imix as well), but for now bumping the kernel ROM seems fine - this chip has 1MB of flash, so giving the kernel some more room to breath ain't bad for now.

**Comment by niklasad1** [comment]
how about a feature_flag (conditional compilation) to enable `BLE` or  `802154`?

**Comment by phil-levis** [comment]
@niklasad1 If we go down this route, we want to do so uniformly across platforms (e.g.. have feature_flag be the standard approach to conditionally include modules). I'd suggest we hold off on that for now, mostly because we want to really check it works everywhere.

**Comment by dcellucci** [comment]
It passes! 

To avoid having to cross the selective compile bridge, I just increased the size of the ROM section in the linker script. To be fair to @niklasad1, the way things are defined now, with the platform described in the nrf52_base section, it's kind of asking for a feature_flag, or at least a flag related to what version of chip is being used. That is, a feature in the radio that the nrf52832 cannot use is being defined in the section that's supposed to be common for all nrf528XX chips. That's just how the platform is set up right now, and I'm not sure if there's a clean fix for that.

Anyway, there is a version of the system than runs either BLE or 802154 on the nordic chips that Travis is happy with. I propose that development on this PR is finished, pending confirmation that things still work on the RF233 (I added that code back in, by the way), and focus on two new PRs, one that focuses on fleshing out the features of the 802154 mode on the 52840 and another that tries to come up with a common HIL for multiprotocol radios.

For the first PR, the features that still need implementation are:
- [ ] ACKnowledgements
- [ ] Proper PPI treatment
- [ ] Incorporation into the bootloader for OTA updates of apps (this one is important for my research)

For the second PR, the TODOs as I see them are as follows:
- [ ] Find some way to combine nrf_radio.rs and ble_radio.rs that makes sense
- [ ] allow MUXing of multiprotocol receive/transmit drivers

What do y'all think?

**Comment by alevy** [comment]
So much discussion to catch up on in such a short time!

I lean towards @niklasad1's idea of using a feature flag since it's not at all clear _how_ to have BLE and 802.15.14 coexist in a real system (yet anyway), and it's much more important to get BLE connections working without the complexity of 802.15.4 coexistence, or dynamic switching. Thinking a bit more about the implications of using feature flags more broadly for now is fine though.

@dcellucci amazing! Let's test the rf233 with this PR (either me and/or @hudson-ayers should be able to to do that fairly quickly in the next few days).

Between the two PRs I think the first is _much_ more important, and it sounds like that's the true for your use case as well. I would probably approach the three features in the first "PR" as three separate ones.

**Comment by hudson-ayers** [comment]
I can test this PR with the RF233 this weekend.

Sent from my iPhone

On Mar 28, 2019, at 9:14 AM, Amit Levy <notifications@github.com<mailto:notifications@github.com>> wrote:


So much discussion to catch up on in such a short time!

I lean towards @niklasad1<https://github.com/niklasad1>'s idea of using a feature flag since it's not at all clear how to have BLE and 802.15.14 coexist in a real system (yet anyway), and it's much more important to get BLE connections working without the complexity of 802.15.4 coexistence, or dynamic switching. Thinking a bit more about the implications of using feature flags more broadly for now is fine though.

@dcellucci<https://github.com/dcellucci> amazing! Let's test the rf233 with this PR (either me and/or @hudson-ayers<https://github.com/hudson-ayers> should be able to to do that fairly quickly in the next few days).

Between the two PRs I think the first is much more important, and it sounds like that's the true for your use case as well. I would probably approach the three features in the first "PR" as three separate ones.

**Comment by hudson-ayers** [comment]
Tested this PR on a couple Imixes and all the UDP tests work as expected! Good job @dcellucci :)

@alevy (so you dont take the time to test it)

**Comment by phil-levis** [comment]
> On Mar 28, 2019, at 6:14 AM, Amit Levy <notifications@github.com> wrote:
> 
> I lean towards @niklasad1 <https://github.com/niklasad1>'s idea of using a feature flag since it's not at all clear how to have BLE and 802.15.14 coexist in a real system (yet anyway), and it's much more important to get BLE connections working without the complexity of 802.15.4 coexistence, or dynamic switching. Thinking a bit more about the implications of using feature flags more broadly for now is fine though.


Meh. I think this will become “just one more new way we do something in some specialized setting which doesn’t quite work generally but kinda works here and is forgotten and then in 6 months people are wondering how it works.”

If the question is compile-time vs. run-time, I agree with compile-time. But the way we’ve done this so far is through instantiation in reset_handler. 

Phil

———————
Philip Levis
Associate Professor, Computer Science and Electrical Engineering
Faculty Director, lab64 Maker Space
Faculty Director, Secure Internet of Things Project
Stanford University
http://csl.stanford.edu/~pal

**Comment by phil-levis** [comment]
Can I bump this? It would be great to get it incorporated.

File: kernel/src/hil/radio.rs (hunk 7 above)
Comment (dcellucci) [reply kernel/src/hil/radio.rs]: I was going off the ble_advertising.rs file in the kernel/src directory. It also defines a `RadioChannel` enum that corresponds to the value being put in the FREQUENCY register.

File: chips/nrf52/src/nrf_radio.rs [hunk 11]
```diff
@@ -0,0 +1,1114 @@
...
+
+        self.transmitting.set(true);
+
+        self.cca_count.set(0);
+        self.cca_be.set(IEEE802154_MIN_BE);
+
+        self.radio_off();
+        self.radio_initialize(self.channel.get());
+
+        //self.enable_interrupts();
```
[the remark this answers is not in the document]
Comment (dcellucci) [reply chips/nrf52/src/nrf_radio.rs]: I'm having a lot of trouble with implementing the `map` call. When I try it, I get an error involving the lifetime of the reference. I guess I'm asking, is there a way to tie the value of the DMA register to the buffers in such a way that the compiler can make sure the DMA is never pointing to an invalid address/incorrect buffer?

[the remark this answers is not in the document]
Comment (alevy) [reply chips/nrf52/src/nrf_radio.rs]: What is this in reference to?

[the remark this answers is not in the document]
Comment (dcellucci) [reply chips/nrf52/src/nrf_radio.rs]: This: 
https://github.com/tock/tock/pull/1237#discussion_r270473471
merged latest master into this branch btw

File: chips/nrf52/src/nrf_radio.rs (hunk 9 above)
Comment (dcellucci) [reply chips/nrf52/src/nrf_radio.rs]: I'm happy to do this, but I'm not exactly sure what needs to be done here. Is there an explainer of the difference between `unwrap`, `expect`, and `map`?

Best, 
DC

Comment (hudson-ayers) [reply chips/nrf52/src/nrf_radio.rs]: Unwrap() will panic if the value in the cell is not there. Given Tock's reliability goals, we generally want to discourage code in the kernel which could lead to panics, as they bring the entire system down. Expect() will also panic if the value is not there, but allows you to pass a custom message, which makes it easier to at least trace where the error originated when debugging. 

Map() makes panic'ing impossible by requiring you to explicitly handle the case of the value not being present. So for the first unwrap in the above code example. you would do something like this (though the None case could vary depending on what should happen if the buffer is missing):
```
if self.transmitting.get() {
            match self.tx_buf.take() {
                    Some(buf) => {
                             self.tx_buf.replace(self.set_dma_ptr(buf));
                    },
                    None => debug!("Error: tx_buf missing"),
             }
// ...
```

Comment (dcellucci) [reply chips/nrf52/src/nrf_radio.rs]: Ah I see. I thought this was more in response to the use of the direct setting of the DMA pointer. Fixed!

**Comment by bradjc** [comment]
This has been merged by #1339, #1340, and #1341 

If adding the BT840 platform as a new board is still of interest please open a new PR for that.

**Pull request closed without merging**